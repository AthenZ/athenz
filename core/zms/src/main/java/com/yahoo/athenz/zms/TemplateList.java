//
// This file generated by rdl 1.5.1. Do not modify!
//

package com.yahoo.athenz.zms;
import java.util.List;
import com.yahoo.rdl.*;

//
// TemplateList - List of template names that is the base struct for server and
// domain templates
//
public class TemplateList {
    public List<String> templateNames;

    public TemplateList setTemplateNames(List<String> templateNames) {
        this.templateNames = templateNames;
        return this;
    }
    public List<String> getTemplateNames() {
        return templateNames;
    }

    @Override
    public boolean equals(Object another) {
        if (this != another) {
            if (another == null || another.getClass() != TemplateList.class) {
                return false;
            }
            TemplateList a = (TemplateList) another;
            if (templateNames == null ? a.templateNames != null : !templateNames.equals(a.templateNames)) {
                return false;
            }
        }
        return true;
    }
}
