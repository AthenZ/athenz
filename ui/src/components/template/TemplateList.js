/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import styled from '@emotion/styled';
import { colors } from '../denali/styles';
import TemplateRow from './TemplateRow';
import Alert from '../denali/Alert';
import { MODAL_TIME_OUT } from '../constants/constants';
import RequestUtils from '../utils/RequestUtils';

const TemplatesSectionDiv = styled.div`
    margin: 20px;
`;

const TemplateTable = styled.table`
    width: 100%;
    border-spacing: 0;
    display: table;
    border-collapse: separate;
    border-color: ${colors.grey600};
`;

const TableHeadStyled = styled.th`
    text-align: ${(props) => props.align};
    border-bottom: 2px solid ${colors.grey500};
    color: ${colors.grey600};
    font-weight: 600;
    font-size: 0.8rem;
    padding-bottom: 5px;
    vertical-align: top;
    text-transform: uppercase;
    padding: 5px 0 5px 15px;
    word-break: break-all;
`;

const TitleDiv = styled.div`
    font: 600 15px HelveticaNeue-Reg, Helvetica, Arial, sans-serif;
    margin-bottom: 20px;
    margin-top: 40px;
`;

export default class TemplateList extends React.Component {
    constructor(props) {
        super(props);
        this.api = props.api;
        this.onCancelUpdateTemplate = this.onCancelUpdateTemplate.bind(this);
        this.reloadTemplates = this.reloadTemplates.bind(this);
        this.showError = this.showError.bind(this);
        this.closeModal = this.closeModal.bind(this);
        let serverTemplateListCopy = [];
        if (props.serverTemplateDetails) {
            serverTemplateListCopy = props.serverTemplateDetails.map((x) => x);

            if (props.domainTemplateDetails) {
                let domainTemplatesArray = props.domainTemplateDetails.map(
                    (template) => {
                        return template.templateName;
                    }
                );

                let domainTemplatesSet = new Set(domainTemplatesArray);
                serverTemplateListCopy = serverTemplateListCopy.filter(
                    (template) => !domainTemplatesSet.has(template.templateName)
                );
            }
        }

        this.state = {
            list: props.domainTemplateDetails || [],
            serverTemplateDetails: serverTemplateListCopy || [],
            successMessage: '',
            errorMessage: '',
            showError: false,
            showSuccess: false,
        };
    }

    showError(errorMessage) {
        this.setState({
            showError: true,
            errorMessage: errorMessage,
        });
    }

    reloadTemplates(successMessage) {
        this.props.api
            .getDomainTemplateDetailsList(this.props.domain)
            .then((data) => {
                let serverTemplateListCopy = [];
                if (this.props.serverTemplateDetails) {
                    serverTemplateListCopy =
                        this.props.serverTemplateDetails.map((x) => x);
                    if (data) {
                        let domainTemplatesArray = data.map((template) => {
                            return template.templateName;
                        });

                        let domainTemplatesSet = new Set(domainTemplatesArray);
                        serverTemplateListCopy = serverTemplateListCopy.filter(
                            (template) =>
                                !domainTemplatesSet.has(template.templateName)
                        );
                    }
                }

                this.setState({
                    list: data,
                    serverTemplateDetails: serverTemplateListCopy || [],
                    showSuccess: true,
                    successMessage,
                });
                // this is to close the success alert
                setTimeout(
                    () =>
                        this.setState({
                            showSuccess: false,
                        }),
                    MODAL_TIME_OUT
                );
            })
            .catch((err) => {
                this.showError(RequestUtils.xhrErrorCheckHelper(err));
            });
    }

    onClickUpdateTemplate() {
        this.setState({
            showUpdate: true,
        });
    }

    closeModal() {
        this.setState({
            showError: false,
            errorMessage: '',
        });
    }

    onCancelUpdateTemplate() {
        this.setState({
            showUpdate: false,
            deleteServiceName: null,
        });
    }

    render() {
        const left = 'left';
        const center = 'center';
        const rows = this.state.list.map((item, i) => {
            let onClickUpdateTemplate = this.onClickUpdateTemplate.bind(this);
            const templateName = item.templateName;
            const templateDesc = item.description;
            let color = '';
            if (i % 2 === 0) {
                color = colors.row;
            }
            let toReturn = [];
            toReturn.push(
                <TemplateRow
                    templateName={templateName}
                    templateDesc={templateDesc}
                    timestamp={item.timestamp}
                    color={color}
                    api={this.api}
                    currentVersion={item.currentVersion}
                    latestVersion={item.latestVersion}
                    keywordsToReplace={item.keywordsToReplace}
                    _csrf={this.props._csrf}
                    key={templateName}
                    onClickUpdateTemplate={onClickUpdateTemplate}
                    domain={this.props.domain}
                    data={this.state.list}
                    onsubmit={this.reloadTemplates}
                    showError={this.showError}
                />
            );
            return toReturn;
        });
        const serverTemplateRows = this.state.serverTemplateDetails.map(
            (item, i) => {
                let onClickUpdateTemplate =
                    this.onClickUpdateTemplate.bind(this);
                const templateName = item.templateName;
                const templateDesc = item.description;
                let color = '';
                if (i % 2 === 0) {
                    color = colors.row;
                }
                let toReturn = [];
                toReturn.push(
                    <TemplateRow
                        templateName={templateName}
                        templateDesc={templateDesc}
                        timestamp={item.timestamp}
                        color={color}
                        api={this.api}
                        currentVersion={'N/A'}
                        latestVersion={item.latestVersion}
                        keywordsToReplace={item.keywordsToReplace}
                        _csrf={this.props._csrf}
                        key={templateName}
                        onClickUpdateTemplate={onClickUpdateTemplate}
                        domain={this.props.domain}
                        data={this.state.serverTemplateDetails}
                        onsubmit={this.reloadTemplates}
                        showError={this.showError}
                    />
                );
                return toReturn;
            }
        );
        return (
            <TemplatesSectionDiv data-testid='template-list'>
                <TitleDiv>Domain Templates</TitleDiv>
                <TemplateTable>
                    <thead>
                        <tr>
                            <TableHeadStyled align={left}>
                                TEMPLATE
                            </TableHeadStyled>
                            <TableHeadStyled align={left}>INFO</TableHeadStyled>
                            <TableHeadStyled align={center}>
                                CURRENT VERSION
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                LATEST VERSION
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                UPDATED DATE
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                UPDATE VERSION
                            </TableHeadStyled>
                        </tr>
                    </thead>
                    <tbody>{rows}</tbody>
                    <TitleDiv>Server Templates</TitleDiv>
                    <thead>
                        <tr>
                            <TableHeadStyled align={left}>
                                TEMPLATE
                            </TableHeadStyled>
                            <TableHeadStyled align={left}>INFO</TableHeadStyled>
                            <TableHeadStyled align={center}>
                                CURRENT VERSION
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                LATEST VERSION
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                UPDATED DATE
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                UPDATE VERSION
                            </TableHeadStyled>
                        </tr>
                    </thead>
                    <tbody>{serverTemplateRows}</tbody>
                </TemplateTable>
                {this.state.showSuccess ? (
                    <Alert
                        isOpen={this.state.showSuccess}
                        title={this.state.successMessage}
                        onClose={this.closeModal}
                        type='success'
                    />
                ) : null}
                {this.state.showError ? (
                    <Alert
                        isOpen={this.state.showError}
                        title={this.state.errorMessage}
                        onClose={this.closeModal}
                        type='danger'
                    />
                ) : null}
            </TemplatesSectionDiv>
        );
    }
}
