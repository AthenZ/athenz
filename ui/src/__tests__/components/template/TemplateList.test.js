/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import TemplateList from '../../../components/template/TemplateList';

describe('TemplateList', () => {
    it('should render', () => {
        const { getByTestId } = render(<TemplateList />);
        const templatelist = getByTestId('template-list');

        expect(templatelist).toMatchSnapshot();
    });

    it('should render with metadata array', () => {
        const left = 'left';
        const center = 'center';
        let toReturn =
            [{
                "templateName": "aws",
                "description": "AWS access template",
                "currentVersion": 4,
                "latestVersion": 1,
                "timestamp": "2020-04-28T00:00:00.000Z",
                "autoUpdate": false
            }];

        const api = {
            updateTemplate: function(params, csrf) {
                return new Promise((resolve, reject) => {
                    resolve(toReturn);
                });
            },
        };

        const { getByTestId } = render(<TemplateList
            left ={left}
            center = {center}
            list={toReturn}
            serverTemplateDetails={toReturn}
            api = {api}
        />);
        const templatelist = getByTestId('template-list');

        expect(templatelist).toMatchSnapshot();
    });

});
