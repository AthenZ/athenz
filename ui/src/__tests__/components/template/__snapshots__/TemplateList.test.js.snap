// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateList should render 1`] = `
.emotion-0 {
  margin: 20px;
}

.emotion-2 {
  font: 600 15px HelveticaNeue-Reg,Helvetica,Arial,sans-serif;
  margin-bottom: 20px;
  margin-top: 40px;
}

.emotion-4 {
  width: 100%;
  border-spacing: 0;
  display: table;
  border-collapse: separate;
  border-color: #9a9a9a;
}

.emotion-6 {
  text-align: left;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  word-break: break-all;
}

.emotion-10 {
  text-align: center;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  word-break: break-all;
}

<div
  class="emotion-0 emotion-1"
  data-testid="template-list"
>
  <div
    class="emotion-2 emotion-3"
  >
    Domain Templates
  </div>
  <table
    class="emotion-4 emotion-5"
  >
    <thead>
      <tr>
        <th
          class="emotion-6 emotion-7"
        >
          TEMPLATE
        </th>
        <th
          class="emotion-6 emotion-7"
        >
          INFO
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          CURRENT VERSION
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          LATEST VERSION
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          UPDATED DATE
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          UPDATE VERSION
        </th>
      </tr>
    </thead>
    <tbody />
    <div
      class="emotion-2 emotion-3"
    >
      Server Templates
    </div>
    <thead>
      <tr>
        <th
          class="emotion-6 emotion-7"
        >
          TEMPLATE
        </th>
        <th
          class="emotion-6 emotion-7"
        >
          INFO
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          CURRENT VERSION
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          LATEST VERSION
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          UPDATED DATE
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          UPDATE VERSION
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`TemplateList should render with metadata array 1`] = `
.emotion-0 {
  margin: 20px;
}

.emotion-2 {
  font: 600 15px HelveticaNeue-Reg,Helvetica,Arial,sans-serif;
  margin-bottom: 20px;
  margin-top: 40px;
}

.emotion-4 {
  width: 100%;
  border-spacing: 0;
  display: table;
  border-collapse: separate;
  border-color: #9a9a9a;
}

.emotion-6 {
  text-align: left;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  word-break: break-all;
}

.emotion-10 {
  text-align: center;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  word-break: break-all;
}

.emotion-32 {
  background-color: #3570f40D;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

.emotion-34 {
  background-color: #3570f40D;
  text-align: center;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

.emotion-36 {
  fill: #188fff;
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-45 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 12px;
  line-height: 1;
  margin: 5px;
  min-width: 70px;
  outline: 0;
  padding: 8px 18px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: transparent;
  box-shadow: 0 0 0 1px #3697f2 inset;
  color: #3697f2;
}

.emotion-45:first-of-type {
  margin-left: 0;
}

.emotion-45:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-45:disabled {
  box-shadow: 0 0 0 1px rgba(48,48,48,0.1) inset;
}

.emotion-45:hover:not(:disabled) {
  background: #d7e2fd;
}

<div
  class="emotion-0 emotion-1"
  data-testid="template-list"
>
  <div
    class="emotion-2 emotion-3"
  >
    Domain Templates
  </div>
  <table
    class="emotion-4 emotion-5"
  >
    <thead>
      <tr>
        <th
          class="emotion-6 emotion-7"
        >
          TEMPLATE
        </th>
        <th
          class="emotion-6 emotion-7"
        >
          INFO
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          CURRENT VERSION
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          LATEST VERSION
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          UPDATED DATE
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          UPDATE VERSION
        </th>
      </tr>
    </thead>
    <tbody />
    <div
      class="emotion-2 emotion-3"
    >
      Server Templates
    </div>
    <thead>
      <tr>
        <th
          class="emotion-6 emotion-7"
        >
          TEMPLATE
        </th>
        <th
          class="emotion-6 emotion-7"
        >
          INFO
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          CURRENT VERSION
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          LATEST VERSION
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          UPDATED DATE
        </th>
        <th
          class="emotion-10 emotion-7"
        >
          UPDATE VERSION
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        data-testid="template-row"
      >
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        >
          aws
        </td>
        <td
          class="emotion-34 emotion-33"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-36"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                information-circle
              </title>
              <path
                d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM512 896c-212.077 0-384-171.923-384-384s171.923-384 384-384c212.077 0 384 171.923 384 384v0c-1.203 211.592-172.408 382.797-383.885 383.999l-0.115 0.001z"
              />
              <path
                d="M512 426.667c-23.564 0-42.667 19.103-42.667 42.667v0 256c0 23.564 19.103 42.667 42.667 42.667s42.667-19.103 42.667-42.667v0-256c0-23.564-19.103-42.667-42.667-42.667v0z"
              />
              <path
                d="M554.667 298.667c0 23.564-19.103 42.667-42.667 42.667s-42.667-19.103-42.667-42.667c0-23.564 19.103-42.667 42.667-42.667s42.667 19.103 42.667 42.667z"
              />
            </svg>
          </span>
        </td>
        <td
          class="emotion-34 emotion-33"
          color="#3570f40D"
        >
          N/A
        </td>
        <td
          class="emotion-34 emotion-33"
          color="#3570f40D"
        >
          1
        </td>
        <td
          class="emotion-34 emotion-33"
          color="#3570f40D"
        >
          2020-04-28T00:00:00.000Z
        </td>
        <td
          class="emotion-34 emotion-33"
          color="#3570f40D"
        >
          <button
            class="emotion-45 denali-button"
            type="submit"
          >
            Onboard
          </button>
          
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
